docs/src/performance_tips.md

```diff
@@ -0,0 +1,62 @@
+# Performance Tips
+
+Here we list various tips that will help users make faster ABMs with Agents.jl.
+These will typically come at the cost of ease of use, or clarity and extend of source code.
+Because otherwise, if they had no downsides, we would have already implemented them in Agents.jl.
+
+Notice that most tips presented here are context-specific. This means if you truly care about performance of your model, because you intend to do massive simulations, it is probably worth it to test all approaches until you conclude which one is the most performant.
+
+Please do read through Julia's own [Performance Tips](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-tips) section as well, as it will help you write performant code in general.
+
+## Avoid `Union`s for multi-agent models
+Due to the way Julia's type system works, and the fact that agents are grouped in a dictionary mapping IDs to agent instances, using multiple types for different agents always creates a performance hit because it leads to type instability.
+
+To avoid this, you can have a single agent type including all properties all kinds should have. You can specify what "kind" of agent it is via including a field `type` or `kind` whose value is a symbol: `:wolf, :sheep, :grass`. Properties that should only belong to one kind of agent could be initialized with a "null" value for the other kinds.
+
+While this will increase slightly the amount of memory used by the model, as all agent instances will contain more data than necessary, the performance gain due to type stability typically makes up for it.
```

Comment:
I don't know if this is the case generally. It could lead to a *lot* of branching in certain cases.

Reply:
>  It could lead to a lot of branching in certain situations.

Only one `if` statement is necessary . `if kind == :wolf; agent_step_wolf!() elseif kind == :sheep; agent_step_sheep!()`, etc.? I should write this down explicitly to improve the docs.